README
KOTS as forked from KOTS2007, which was a reconstruction of KOTS2000, which...


changelog.txt
Version 2.0.2 - 2021-07-04 - Kold
CHANGE: Modified hook to be a laser instead of the classic CTF grapple. Added ability to change hook color.
CHANGE: Added chaingun ammo convert.
CHANGE: Changed all tabs to spaces to enforce consistent style.

Version 2.0.1 - 2021-07-04 - Kold
BUGFIX: Forced gamedir to kots2020 for consistent behavior across clients, since Q2PRO doesn't populate gamedir
CHANGE: Nerfed SSG by reducing pellet increase in 2.0.0
CHANGE: Modified reunion splash

Version 2.0.0 - 2021-07-02 - Kold
BUGFIX: Fixed a typo in kots_db.sql causing syntax errors when imported.
CHANGE: Buffed SSG by adding more pellets at higher levels
CHANGE: Buffed MG by reducing bullet spread at higher levels
CHANGE: Buffed HB by increasing damage and tracer speed at higher levels
CHANGE: Nerfed bide by creating a particle effect while charging to warn other players
CHANGE: Buffed dex by making hook silent at level 7
CHANGE: Buffed dex by changing hook cube cost at level 7 from 4 to 2
CHANGE: Increased exp rate and credit pickups by a factor of 10.
CHANGE: Nerfed tech armor regen: now only applied when tech is mastered.
CHANGE: Nerfed tech armor regen from 4/sec to 2/sec.
CHANGE: Tech armor resist used to be level 3: 1.4; level 5: 1.75; level 7: 2.25. It is now level 3: 1.5; level 5+: 2.
CHANGE: Spiral can only be trained once every three levels until one hits level 30.
CHANGE: Modified the text that is displayed when you login to advertise the mod page and give a reunion feel.

Version 1.1.0 - 2010-08-15 - SWB
BUGFIX: Fixed a bug where the wrong text printed out for sv dbhostname.
BUGFIX: Fixed a bug with the wrong dbhostname variable being used.
BUGFIX: Fixed some Linux build issues with some code in kots_update.c (Don't remember where this came from but it wasn't used).
CHANGE: Nerfed fly to limit it to 3 seconds of flying time with a cooldown and recharge period.
CHANGE: Nerfed the hook to limit it to only hold on for 5 seconds with a half second cooldown.
CHANGE: Removed the ability to train weapons above level 6.
CHANGE: Nerfed grenade launcher level 10 effect to only last 3 seconds or 1.5 with strength reduction.
CHANGE: Nerfed hyperblaster level 10 effect to only have a 50% chance to happen.
CHANGE: Reduced tech 3 to resistance 1.4
CHANGE: Added tech 5 resistance of 1.5
CHANGE: Capped munition and vitality to 1 per 3 levels until you reach level 10
CHANGE: Added respec function for beginners: Upon reaching level 10 you have the option of relocating 5 player points.

Changes provided thanks to Corrupted:
Changed leveling a bit to have people go through the midlevels faster:
   Level 8: 60k
   Level 9: 100k
   Level 10: 150k
   Level 11: 200k
   Level 12: 275k
   Level 13: 350k
   Level 14: 425k
   Level 15: 500k
   Level 16: 600k

CHANGE: Nerfed vithealth: Adds 35 life now
CHANGE: Nerfed spiral growth: 100 damage initially, 100 damage per level, can't be trained before level 6
CHANGE: Nerfed weapon knock: str7 from 160 to 250, str 5 from 250 to 400 (didn't test yet)
CHANGE: Nerfed monsters exp/credits significantly (changes mostly aimed at people abusing stupid small maps like 2box4)
CHANGE: Nerfed railgun (lvl 5: 200 dmg, lvl 6: 250 dmg)
CHANGE: Buffed sprees (to 2x, 3x, 5x bonus)
CHANGE: Nerfed hand grenade (lvl 6: 350 dmg, lvl 10: 400 dmg, 1000 speed)
CHANGE: Nerfed hook speed/increased cubecost

Version 1.0.10 - 2009-01-24 - SWB
Cleaned up code and added variables for people to host servers using their own database.

Version 1.0.9 - 2008-03-26 - SWB
BUGFIX: Karma level 2 was not allowing pickup beyond max anymore
BUGFIX: Fixed a problem where your own packs weren't counting towards your total credits in kots_stats
BUGFIX: Fixed a false positive for dummy killing that can happen from killing monsters
BUGFIX: A crash would occur when a player disconnected before login completed

Version 1.0.8 - 2008-03-23 - SWB
BUGFIX: Fixed the lack of line breaks on kots_stats help
BUGFIX: Fixed a few database connection issues
BUGFIX: Fixed a bug that could cause crashes in 1.0.7 when logging information
BUGFIX: Fixed a bug that could crash the server during shutdown
BUGFIX: The thump sound when on your spree was not always indexed each map so you might not always hear it

Version 1.0.7 - 2008-03-23 - SWB
BUGFIX: Certain server cvars could be changed during the map that shouldn't (they would be reset next map though)
BUGFIX: Fixed a bug with chaingun accuracy
BUGFIX: Fixed a bug with kots health not curing poison
BUGFIX: Fixed a bug with bide death explosions that was causing illegitimate 2fers
BUGFIX: Fixed a bug with the HG short fuse setting

game/g_save.c
/*
Copyright (C) 1997-2001 Id Software, Inc.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

*/

#include "g_local.h"

//SWB
#include "kots_cvars.h"
#include "kots_monster.h"


#define Function(f) {#f, f}

mmove_t mmove_reloc;

field_t fields[] = {
    {"classname", FOFS(classname), F_LSTRING},
    {"model", FOFS(model), F_LSTRING},
    {"spawnflags", FOFS(spawnflags), F_INT},
    {"speed", FOFS(speed), F_FLOAT},
    {"accel", FOFS(accel), F_FLOAT},
    {"decel", FOFS(decel), F_FLOAT},
    {"target", FOFS(target), F_LSTRING},
    {"targetname", FOFS(targetname), F_LSTRING},
    {"pathtarget", FOFS(pathtarget), F_LSTRING},
    {"deathtarget", FOFS(deathtarget), F_LSTRING},
    {"killtarget", FOFS(killtarget), F_LSTRING},
    {"combattarget", FOFS(combattarget), F_LSTRING},
    {"message", FOFS(message), F_LSTRING},
    {"team", FOFS(team), F_LSTRING},
    {"wait", FOFS(wait), F_FLOAT},
    {"delay", FOFS(delay), F_FLOAT},
    {"random", FOFS(random), F_FLOAT},
    {"move_origin", FOFS(move_origin), F_VECTOR},
    {"move_angles", FOFS(move_angles), F_VECTOR},
    {"style", FOFS(style), F_INT},
    {"count", FOFS(count), F_INT},
    {"health", FOFS(health), F_INT},
    {"sounds", FOFS(sounds), F_INT},
    {"light", 0, F_IGNORE},
    {"dmg", FOFS(dmg), F_INT},
    {"mass", FOFS(mass), F_INT},
    {"volume", FOFS(volume), F_FLOAT},
    {"attenuation", FOFS(attenuation), F_FLOAT},
    {"map", FOFS(map), F_LSTRING},
    {"origin", FOFS(s.origin), F_VECTOR},
    {"angles", FOFS(s.angles), F_VECTOR},
    {"angle", FOFS(s.angles), F_ANGLEHACK},

    {"goalentity", FOFS(goalentity), F_EDICT, FFL_NOSPAWN},
    {"movetarget", FOFS(movetarget), F_EDICT, FFL_NOSPAWN},
    {"enemy", FOFS(enemy), F_EDICT, FFL_NOSPAWN},
    {"oldenemy", FOFS(oldenemy), F_EDICT, FFL_NOSPAWN},
    {"activator", FOFS(activator), F_EDICT, FFL_NOSPAWN},
    {"groundentity", FOFS(groundentity), F_EDICT, FFL_NOSPAWN},
    {"teamchain", FOFS(teamchain), F_EDICT, FFL_NOSPAWN},
    {"teammaster", FOFS(teammaster), F_EDICT, FFL_NOSPAWN},
    {"owner", FOFS(owner), F_EDICT, FFL_NOSPAWN},
    {"mynoise", FOFS(mynoise), F_EDICT, FFL_NOSPAWN},
    {"mynoise2", FOFS(mynoise2), F_EDICT, FFL_NOSPAWN},
    {"target_ent", FOFS(target_ent), F_EDICT, FFL_NOSPAWN},
    {"chain", FOFS(chain), F_EDICT, FFL_NOSPAWN},

    {"prethink", FOFS(prethink), F_FUNCTION, FFL_NOSPAWN},
    {"think", FOFS(think), F_FUNCTION, FFL_NOSPAWN},
    {"blocked", FOFS(blocked), F_FUNCTION, FFL_NOSPAWN},
    {"touch", FOFS(touch), F_FUNCTION, FFL_NOSPAWN},
    {"use", FOFS(use), F_FUNCTION, FFL_NOSPAWN},
    {"pain", FOFS(pain), F_FUNCTION, FFL_NOSPAWN},
    {"die", FOFS(die), F_FUNCTION, FFL_NOSPAWN},

    {"stand", FOFS(monsterinfo.stand), F_FUNCTION, FFL_NOSPAWN},
    {"idle", FOFS(monsterinfo.idle), F_FUNCTION, FFL_NOSPAWN},
    {"search", FOFS(monsterinfo.search), F_FUNCTION, FFL_NOSPAWN},
    {"walk", FOFS(monsterinfo.walk), F_FUNCTION, FFL_NOSPAWN},
    {"run", FOFS(monsterinfo.run), F_FUNCTION, FFL_NOSPAWN},
    {"dodge", FOFS(monsterinfo.dodge), F_FUNCTION, FFL_NOSPAWN},
    {"attack", FOFS(monsterinfo.attack), F_FUNCTION, FFL_NOSPAWN},
    {"melee", FOFS(monsterinfo.melee), F_FUNCTION, FFL_NOSPAWN},
    {"sight", FOFS(monsterinfo.sight), F_FUNCTION, FFL_NOSPAWN},
    {"checkattack", FOFS(monsterinfo.checkattack), F_FUNCTION, FFL_NOSPAWN},
    {"currentmove", FOFS(monsterinfo.currentmove), F_MMOVE, FFL_NOSPAWN},

    {"endfunc", FOFS(moveinfo.endfunc), F_FUNCTION, FFL_NOSPAWN},

    // temp spawn vars -- only valid when the spawn function is called
    {"lip", STOFS(lip), F_INT, FFL_SPAWNTEMP},
    {"distance", STOFS(distance), F_INT, FFL_SPAWNTEMP},
    {"height", STOFS(height), F_INT, FFL_SPAWNTEMP},
    {"noise", STOFS(noise), F_LSTRING, FFL_SPAWNTEMP},
    {"pausetime", STOFS(pausetime), F_FLOAT, FFL_SPAWNTEMP},
    {"item", STOFS(item), F_LSTRING, FFL_SPAWNTEMP},

//need for item field in edict struct, FFL_SPAWNTEMP item will be skipped on saves
    {"item", FOFS(item), F_ITEM},

    {"gravity", STOFS(gravity), F_LSTRING, FFL_SPAWNTEMP},
    {"sky", STOFS(sky), F_LSTRING, FFL_SPAWNTEMP},
    {"skyrotate", STOFS(skyrotate), F_FLOAT, FFL_SPAWNTEMP},
    {"skyaxis", STOFS(skyaxis), F_VECTOR, FFL_SPAWNTEMP},
    {"minyaw", STOFS(minyaw), F_FLOAT, FFL_SPAWNTEMP},
    {"maxyaw", STOFS(maxyaw), F_FLOAT, FFL_SPAWNTEMP},
    {"minpitch", STOFS(minpitch), F_FLOAT, FFL_SPAWNTEMP},
    {"maxpitch", STOFS(maxpitch), F_FLOAT, FFL_SPAWNTEMP},
    {"nextmap", STOFS(nextmap), F_LSTRING, FFL_SPAWNTEMP},

    {0, 0, 0, 0}

};

field_t     levelfields[] =
{
    {"changemap", LLOFS(changemap), F_LSTRING},

    {"sight_client", LLOFS(sight_client), F_EDICT},
    {"sight_entity", LLOFS(sight_entity), F_EDICT},
    {"sound_entity", LLOFS(sound_entity), F_EDICT},
    {"sound2_entity", LLOFS(sound2_entity), F_EDICT},

    {NULL, 0, F_INT}
};

field_t     clientfields[] =
{
    {"pers.weapon", CLOFS(pers.weapon), F_ITEM},
    {"pers.lastweapon", CLOFS(pers.lastweapon), F_ITEM},
    {"newweapon", CLOFS(newweapon), F_ITEM},

    {NULL, 0, F_INT}
};


/*
============
InitGame

This will be called when the dll is first loaded, which
only happens when a new game is started or a save game
is loaded.
============
*/
void InitGame (void)
{
    //Kots_UpdateGetModulePath();
    gi.dprintf ("==== InitGame ====\n");

    gun_x = gi.cvar ("gun_x", "0", 0);
    gun_y = gi.cvar ("gun_y", "0", 0);
    gun_z = gi.cvar ("gun_z", "0", 0);

    //FIXME: sv_ prefix is wrong for these
    sv_rollspeed = gi.cvar ("sv_rollspeed", "200", 0);
    sv_rollangle = gi.cvar ("sv_rollangle", "2", 0);
    sv_maxvelocity = gi.cvar ("sv_maxvelocity", "2000", 0);
    sv_gravity = gi.cvar ("sv_gravity", "800", 0);

    // noset vars
    dedicated = gi.cvar ("dedicated", "0", CVAR_NOSET);

    // latched vars
    //SWB - changed to nosetm
    sv_cheats = gi.cvar ("cheats", "0", CVAR_SERVERINFO | CVAR_NOSET);
    gi.cvar ("gamename", GAMEVERSION , CVAR_SERVERINFO | CVAR_NOSET);
    gi.cvar ("gamedate", __DATE__ , CVAR_SERVERINFO | CVAR_NOSET);

    maxclients = gi.cvar ("maxclients", "4", CVAR_SERVERINFO | CVAR_LATCH);
    maxspectators = gi.cvar ("maxspectators", "4", CVAR_SERVERINFO | CVAR_LATCH);
    deathmatch = gi.cvar ("deathmatch", "1", CVAR_NOSET | CVAR_LATCH);
    coop = gi.cvar ("coop", "0", CVAR_NOSET | CVAR_LATCH);
    skill = gi.cvar ("skill", "5", CVAR_NOSET | CVAR_LATCH);
    maxentities = gi.cvar ("maxentities", "1024", CVAR_LATCH);

    // change anytime vars
    dmflags = gi.cvar ("dmflags", "0", CVAR_SERVERINFO | CVAR_NOSET | CVAR_LATCH);
    fraglimit = gi.cvar ("fraglimit", "0", CVAR_SERVERINFO |CVAR_NOSET |  CVAR_LATCH);
    timelimit = gi.cvar ("timelimit", "0", CVAR_SERVERINFO | CVAR_NOSET | CVAR_LATCH);
    password = gi.cvar ("password", "", CVAR_USERINFO);
    spectator_password = gi.cvar ("spectator_password", "", CVAR_USERINFO);
    needpass = gi.cvar ("needpass", "0", CVAR_SERVERINFO);
    filterban = gi.cvar ("filterban", "1", 0);

    g_select_empty = gi.cvar ("g_select_empty", "0", CVAR_ARCHIVE);

    run_pitch = gi.cvar ("run_pitch", "0.002", 0);
    run_roll = gi.cvar ("run_roll", "0.005", 0);
    bob_up  = gi.cvar ("bob_up", "0.005", 0);
    bob_pitch = gi.cvar ("bob_pitch", "0.002", 0);
    bob_roll = gi.cvar ("bob_roll", "0.002", 0);

    //SWB
    kots_pass = gi.cvar("kots_pass", "", CVAR_USERINFO);
    kots_nas = gi.cvar("nas", "0", CVAR_USERINFO);
    kots_maplist = gi.cvar("maplist", "", CVAR_SERVERINFO);
    kots_maprotate = gi.cvar("maprotate", "1", CVAR_SERVERINFO);
    kots_public = gi.cvar("public", "1", CVAR_SERVERINFO);
    kots_motd = gi.cvar("motd", "motd.txt", CVAR_SERVERINFO);
    kots_prevmaps = gi.cvar("prevmaps", "5", CVAR_SERVERINFO);
    gamever = gi.cvar("gamever", "", CVAR_SERVERINFO | CVAR_NOSET);
    // Kold - force this to kots2020 to enforce consistent behavior,
    //        since some clients do not populate it (Q2PRO)
    gamedir = gi.cvar("gamedir", "kots2020", CVAR_NOSET);
    hostname = gi.cvar("hostname", "", CVAR_SERVERINFO);
    mapname = gi.cvar("mapname", "", CVAR_SERVERINFO | CVAR_NOSET);
    port = gi.cvar("port", "", CVAR_SERVERINFO | CVAR_NOSET);

    //force the version to get updated
    gi.cvar_forceset(gamever->name, KOTS_VERSION);

    //SWB - announce the version
    gi.dprintf("Starting %s\n", gamever->string);

    // flood control
    flood_msgs = gi.cvar ("flood_msgs", "4", 0);
    flood_persecond = gi.cvar ("flood_persecond", "4", 0);
    flood_waitdelay = gi.cvar ("flood_waitdelay", "10", 0);

    // dm map list
    sv_maplist = gi.cvar ("sv_maplist", "", 0);

    // items
    InitItems ();

    Com_sprintf (game.helpmessage1, sizeof(game.helpmessage1), "");

    Com_sprintf (game.helpmessage2, sizeof(game.helpmessage2), "");

    // initialize all entities for this game
    game.maxentities = maxentities->value;
    g_edicts =  gi.TagMalloc (game.maxentities * sizeof(g_edicts[0]), TAG_GAME);
    globals.edicts = g_edicts;
    globals.max_edicts = game.maxentities;

    // initialize all clients for this game
    game.maxclients = maxclients->value;
    game.clients = gi.TagMalloc (game.maxclients * sizeof(game.clients[0]), TAG_GAME);
    globals.num_edicts = game.maxclients+1;

#ifdef KOTS_PVP_ONLY
    KOTS_MONSTER_VOTE = KOTS_MONSTERS_OFF;
#else
    KOTS_MONSTER_VOTE = KOTS_MONSTERS_MIXED;
#endif
}

//=========================================================

void WriteField1 (FILE *f, field_t *field, byte *base)
{
    void        *p;
    int         len;
    int         index;

    if (field->flags & FFL_SPAWNTEMP)
        return;

    p = (void *)(base + field->ofs);
    switch (field->type)
    {
    case F_INT:
    case F_FLOAT:
    case F_ANGLEHACK:
    case F_VECTOR:
    case F_IGNORE:
        break;

    case F_LSTRING:
    case F_GSTRING:
        if ( *(char **)p )
            len = strlen(*(char **)p) + 1;
        else
            len = 0;
        *(int *)p = len;
        break;
    case F_EDICT:
        if ( *(edict_t **)p == NULL)
            index = -1;
        else
            index = *(edict_t **)p - g_edicts;
        *(int *)p = index;
        break;
    case F_CLIENT:
        if ( *(gclient_t **)p == NULL)
            index = -1;
        else
            index = *(gclient_t **)p - game.clients;
        *(int *)p = index;
        break;
    case F_ITEM:
        if ( *(edict_t **)p == NULL)
            index = -1;
        else
            index = *(gitem_t **)p - itemlist;
        *(int *)p = index;
        break;

    //relative to code segment
    case F_FUNCTION:
        if (*(byte **)p == NULL)
            index = 0;
        else
            index = *(byte **)p - ((byte *)InitGame);
        *(int *)p = index;
        break;

    //relative to data segment
    case F_MMOVE:
        if (*(byte **)p == NULL)
            index = 0;
        else
            index = *(byte **)p - (byte *)&mmove_reloc;
        *(int *)p = index;
        break;

    default:
        gi.error ("WriteEdict: unknown field type");
    }
}


void WriteField2 (FILE *f, field_t *field, byte *base)
{
    int         len;
    void        *p;

    if (field->flags & FFL_SPAWNTEMP)
        return;

    p = (void *)(base + field->ofs);
    switch (field->type)
    {
    case F_LSTRING:
        if ( *(char **)p )
        {
            len = strlen(*(char **)p) + 1;
            fwrite (*(char **)p, len, 1, f);
        }
        break;
    }
}

void ReadField (FILE *f, field_t *field, byte *base)
{
    void        *p;
    int         len;
    int         index;

    if (field->flags & FFL_SPAWNTEMP)
        return;

    p = (void *)(base + field->ofs);
    switch (field->type)
    {
    case F_INT:
    case F_FLOAT:
    case F_ANGLEHACK:
    case F_VECTOR:
    case F_IGNORE:
        break;

    case F_LSTRING:
        len = *(int *)p;
        if (!len)
            *(char **)p = NULL;
        else
        {
            *(char **)p = gi.TagMalloc (len, TAG_LEVEL);
            fread (*(char **)p, len, 1, f);
        }
        break;
    case F_EDICT:
        index = *(int *)p;
        if ( index == -1 )
            *(edict_t **)p = NULL;
        else
            *(edict_t **)p = &g_edicts[index];
        break;
    case F_CLIENT:
        index = *(int *)p;
        if ( index == -1 )
            *(gclient_t **)p = NULL;
        else
            *(gclient_t **)p = &game.clients[index];
        break;
    case F_ITEM:
        index = *(int *)p;
        if ( index == -1 )
            *(gitem_t **)p = NULL;
        else
            *(gitem_t **)p = &itemlist[index];
        break;

    //relative to code segment
    case F_FUNCTION:
        index = *(int *)p;
        if ( index == 0 )
            *(byte **)p = NULL;
        else
            *(byte **)p = ((byte *)InitGame) + index;
        break;

    //relative to data segment
    case F_MMOVE:
        index = *(int *)p;
        if (index == 0)
            *(byte **)p = NULL;
        else
            *(byte **)p = (byte *)&mmove_reloc + index;
        break;

    default:
        gi.error ("ReadEdict: unknown field type");
    }
}

//=========================================================

/*
==============
WriteClient

All pointer variables (except function pointers) must be handled specially.
==============
*/
void WriteClient (FILE *f, gclient_t *client)
{
    field_t     *field;
    gclient_t   temp;

    // all of the ints, floats, and vectors stay as they are
    temp = *client;

    // change the pointers to lengths or indexes
    for (field=clientfields ; field->name ; field++)
    {
        WriteField1 (f, field, (byte *)&temp);
    }

    // write the block
    fwrite (&temp, sizeof(temp), 1, f);

    // now write any allocated data following the edict
    for (field=clientfields ; field->name ; field++)
    {
        WriteField2 (f, field, (byte *)client);
    }
}

/*
==============
ReadClient

All pointer variables (except function pointers) must be handled specially.
==============
*/
void ReadClient (FILE *f, gclient_t *client)
{
    field_t     *field;

    fread (client, sizeof(*client), 1, f);

    for (field=clientfields ; field->name ; field++)
    {
        ReadField (f, field, (byte *)client);
    }
}

/*
============
WriteGame

This will be called whenever the game goes to a new level,
and when the user explicitly saves the game.

Game information include cross level data, like multi level
triggers, help computer info, and all client states.

A single player death will automatically restore from the
last save position.
============
*/
void WriteGame (char *filename, qboolean autosave)
{
    FILE    *f;
    int     i;
    char    str[16];

    if (!autosave)
        SaveClientData ();

    f = fopen (filename, "wb");
    if (!f)
        gi.error ("Couldn't open %s", filename);

    memset (str, 0, sizeof(str));
    strcpy (str, __DATE__);
    fwrite (str, sizeof(str), 1, f);

    game.autosaved = autosave;
    fwrite (&game, sizeof(game), 1, f);
    game.autosaved = false;

    for (i=0 ; i<game.maxclients ; i++)
        WriteClient (f, &game.clients[i]);

    fclose (f);
}

void ReadGame (char *filename)
{
    FILE    *f;
    int     i;
    char    str[16];

    gi.FreeTags (TAG_GAME);

    f = fopen (filename, "rb");
    if (!f)
        gi.error ("Couldn't open %s", filename);

    fread (str, sizeof(str), 1, f);
    if (strcmp (str, __DATE__))
    {
        fclose (f);
        gi.error ("Savegame from an older version.\n");
    }

    g_edicts =  gi.TagMalloc (game.maxentities * sizeof(g_edicts[0]), TAG_GAME);
    globals.edicts = g_edicts;

    fread (&game, sizeof(game), 1, f);
    game.clients = gi.TagMalloc (game.maxclients * sizeof(game.clients[0]), TAG_GAME);
    for (i=0 ; i<game.maxclients ; i++)
        ReadClient (f, &game.clients[i]);

    fclose (f);
}

//==========================================================


/*
==============
WriteEdict

All pointer variables (except function pointers) must be handled specially.
==============
*/
void WriteEdict (FILE *f, edict_t *ent)
{
    field_t     *field;
    edict_t     temp;

    // all of the ints, floats, and vectors stay as they are
    temp = *ent;

    // change the pointers to lengths or indexes
    for (field=fields ; field->name ; field++)
    {
        WriteField1 (f, field, (byte *)&temp);
    }

    // write the block
    fwrite (&temp, sizeof(temp), 1, f);

    // now write any allocated data following the edict
    for (field=fields ; field->name ; field++)
    {
        WriteField2 (f, field, (byte *)ent);
    }

}

/*
==============
WriteLevelLocals

All pointer variables (except function pointers) must be handled specially.
==============
*/
void WriteLevelLocals (FILE *f)
{
    field_t     *field;
    level_locals_t      temp;

    // all of the ints, floats, and vectors stay as they are
    temp = level;

    // change the pointers to lengths or indexes
    for (field=levelfields ; field->name ; field++)
    {
        WriteField1 (f, field, (byte *)&temp);
    }

    // write the block
    fwrite (&temp, sizeof(temp), 1, f);

    // now write any allocated data following the edict
    for (field=levelfields ; field->name ; field++)
    {
        WriteField2 (f, field, (byte *)&level);
    }
}


/*
==============
ReadEdict

All pointer variables (except function pointers) must be handled specially.
==============
*/
void ReadEdict (FILE *f, edict_t *ent)
{
    field_t     *field;

    fread (ent, sizeof(*ent), 1, f);

    for (field=fields ; field->name ; field++)
    {
        ReadField (f, field, (byte *)ent);
    }
}

/*
==============
ReadLevelLocals

All pointer variables (except function pointers) must be handled specially.
==============
*/
void ReadLevelLocals (FILE *f)
{
    field_t     *field;

    fread (&level, sizeof(level), 1, f);

    for (field=levelfields ; field->name ; field++)
    {
        ReadField (f, field, (byte *)&level);
    }
}

/*
=================
WriteLevel

=================
*/
void WriteLevel (char *filename)
{
    int     i;
    edict_t *ent;
    FILE    *f;
    void    *base;

    f = fopen (filename, "wb");
    if (!f)
        gi.error ("Couldn't open %s", filename);

    // write out edict size for checking
    i = sizeof(edict_t);
    fwrite (&i, sizeof(i), 1, f);

    // write out a function pointer for checking
    base = (void *)InitGame;
    fwrite (&base, sizeof(base), 1, f);

    // write out level_locals_t
    WriteLevelLocals (f);

    // write out all the entities
    for (i=0 ; i<globals.num_edicts ; i++)
    {
        ent = &g_edicts[i];
        if (!ent->inuse)
            continue;
        fwrite (&i, sizeof(i), 1, f);
        WriteEdict (f, ent);
    }
    i = -1;
    fwrite (&i, sizeof(i), 1, f);

    fclose (f);
}


/*
=================
ReadLevel

SpawnEntities will allready have been called on the
level the same way it was when the level was saved.

That is necessary to get the baselines
set up identically.

The server will have cleared all of the world links before
calling ReadLevel.

No clients are connected yet.
=================
*/
void ReadLevel (char *filename)
{
    int     entnum;
    FILE    *f;
    int     i;
    void    *base;
    edict_t *ent;

    f = fopen (filename, "rb");
    if (!f)
        gi.error ("Couldn't open %s", filename);

    // free any dynamic memory allocated by loading the level
    // base state
    gi.FreeTags (TAG_LEVEL);

    // wipe all the entities
    memset (g_edicts, 0, game.maxentities*sizeof(g_edicts[0]));
    globals.num_edicts = maxclients->value+1;

    // check edict size
    fread (&i, sizeof(i), 1, f);
    if (i != sizeof(edict_t))
    {
        fclose (f);
        gi.error ("ReadLevel: mismatched edict size");
    }

    // check function pointer base address
    fread (&base, sizeof(base), 1, f);
#ifdef _WIN32
    if (base != (void *)InitGame)
    {
        fclose (f);
        gi.error ("ReadLevel: function pointers have moved");
    }
#else
    gi.dprintf("Function offsets %d\n", ((byte *)base) - ((byte *)InitGame));
#endif

    // load the level locals
    ReadLevelLocals (f);

    // load all the entities
    while (1)
    {
        if (fread (&entnum, sizeof(entnum), 1, f) != 1)
        {
            fclose (f);
            gi.error ("ReadLevel: failed to read entnum");
        }
        if (entnum == -1)
            break;
        if (entnum >= globals.num_edicts)
            globals.num_edicts = entnum+1;

        ent = &g_edicts[entnum];
        ReadEdict (f, ent);

        // let the server rebuild world links for this ent
        memset (&ent->area, 0, sizeof(ent->area));
        gi.linkentity (ent);
    }

    fclose (f);

    // mark all clients as unconnected
    for (i=0 ; i<maxclients->value ; i++)
    {
        ent = &g_edicts[i+1];
        ent->client = game.clients + i;
        ent->client->pers.connected = false;
    }

    // do any load time things at this point
    for (i=0 ; i<globals.num_edicts ; i++)
    {
        ent = &g_edicts[i];

        if (!ent->inuse)
            continue;

        // fire any cross-level triggers
        if (ent->classname)
            if (strcmp(ent->classname, "target_crosslevel_target") == 0)
                ent->nextthink = level.time + ent->delay;
    }
}
